src/outputs/rcp.c

```diff
@@ -0,0 +1,1484 @@
+      {
+	s->state = RCP_STATE_DISCONNECTED;
+	return -1;
+      }
+
+    s->state = next_state;
+
+    return 0;
+}
+
+/* Returns:
+ *   -1  - failure of some kind on link
+ *    0  - recv'd data
+ *
+ * reads data
+ */
+static int
+rcp_recv(struct rcp_session *s)
+{
+  /* The RCP responses are of finite size so we can
+   * limit the input buf
+   */
+  ssize_t recvd;
+  const size_t  avail = RCP_RESP_BUF_SIZE - (s->respptr - s->respbuf);
+
+  if (avail == 0)
+    {
+      DPRINTF(E_WARN, L_RCP, "protocol BUG, cmd buf (%d) exhausted %ld state %d\n", RCP_RESP_BUF_SIZE, s->device->id, s->state);
+
+      s->state = RCP_STATE_FAILED;
+      return -1;
+    }
+
+  recvd = read(s->sock, s->respptr, avail);
+
+  DPRINTF(E_DBG,  L_RCP, "%ld state %d recv'd %ld bytes '%s'\n", s->device->id, s->state, recvd, s->respptr);
+  if (recvd <= 0)
+    {
+      DPRINTF(E_LOG,  L_RCP, "Failed to read response from '%s' - %s\n", s->devname, strerror(recvd == 0 ? ECONNRESET : errno));
+      s->state = RCP_STATE_DISCONNECTED;
+      return -1;
+    }
+
+  s->respptr += recvd;
+  return 0;
+}
+
+/* Returns non-null ptr to a single response
+ * Roku can send multiple responses in packet
+ */
+static const char *
+rcp_state_1resp(char *resp, struct rcp_session *s)
+{
+  char *p, *q;
+  int len;
+  char *ret = NULL;
+
+  // find termination sequence of '\r\n' and adjust the incoming write position
+  // s->respptr, accordingly
+
+  // verify response termination sequence of '\r\n' but expected at least cmd + ':'
+  // ie min response is 'A: OK\r\n'
+  if (s->respptr == s->respbuf || s->respptr - s->respbuf < 6)
+    return NULL;
+
+  p = s->respbuf;
+  q = p+1;
+  while (q < s->respptr)
+    {
+      if (*p == '\r' && *q == '\n')
+      {
+	  len = q+1 - s->respbuf;
+	  memcpy(resp, s->respbuf, len);
+	  resp[len] = '\0';
+
+	  // now slide the rest of the s->respbuf to begining
+	  memmove(s->respbuf, s->respbuf+len, s->respptr - s->respbuf - len);
+	  s->respptr -= len;
+	  memset(s->respptr, 0, len);
+
+	  ret = &resp[0];
+      }
+      ++p;
+      ++q;
+
+      if (ret)
+	break;
+   }
+  return ret;
+}
+
+/* ---------------------------- SESSION HANDLING ---------------------------- */
+
+void
+rcp_disconnect(int fd)
+{
+  /* no more receptions */
+  shutdown(fd, SHUT_RDWR);
+  close(fd);
+}
+
+static int 
+rcp_session_shutdown_clear(struct rcp_session* s)
+{
+  const char *err;
+  char cmd[RCP_RESP_BUF_SIZE+1] = { 0 };
+  fd_set fdset;
+  struct timeval clear_timeout = { 15, 0 };
+  const char *p;
+  int ret;
+
+  memset(s->respbuf, 0, RCP_RESP_BUF_SIZE);
+  s->respptr = s->respbuf;
+  if (s->clear_on_close_timeout)
+    clear_timeout.tv_sec = s->clear_on_close_timeout;
+
+  DPRINTF(E_INFO, L_RCP, "Clearing on shutdown (timeout %ld sec) '%s' at %s\n", clear_timeout.tv_sec, s->devname, s->address);
+
+  rcp_send(s, RCP_STATE_SHUTDOWN_STOPPED, NULL);
+  rcp_status(s);
+
+  while (s->state != RCP_STATE_FAILED && s->state != RCP_STATE_SHUTDOWN_SERVER_DISCONNECT)
```

Comment:
Is this a wait loop? An output module shouldn't do anything that blocks, since that blocks the player too. So everything needs to be done async.

Reply:
in the session shutdown, is it still acceptable to leave it up to the registered `listen_cb` that handles the normal response?  to deal with slow/no response from the remote device, add a timeout cb to ignore the remote device and forcibly drop the connection?